benches/hvm_cpu.rs

```diff
@@ -29,57 +35,75 @@ fn load_from_lang(file: &str, size: usize, replace: Option<(&str, &str)>) -> (ru
-  }
+fn run_programs_dir(c: &mut Criterion) {
+  let root = PathBuf::from(format!("{}/benches/programs", env!("CARGO_MANIFEST_DIR")));
+  run_dir(&root, None, c);
 }
 
-fn binary_counter_benchmark(c: &mut Criterion) {
-  let mut group = c.benchmark_group("binary-counter");
-  // Decrements a BitString until it is zero
-  for n in [12] {
-    let (book, net) = load_from_core("./benches/programs/dec_bits.hvmc", 16 << n, Some(("{n}", &n.to_string())));
-    group.bench_with_input(criterion::BenchmarkId::new("single", n), &n, |b, &_n| {
-      b.iter_batched(|| net.clone(), |mut net| black_box(net.normal(&book)), criterion::BatchSize::SmallInput);
-    });
-  }
-  // Decrements 2^N BitStrings until they reach zero (ex3)
-  for n in [6] {
-    let (book, net) = load_from_core("./benches/programs/dec_bits_tree.hvmc", 128 << n, Some(("{n}", &n.to_string())));
-    group.bench_with_input(criterion::BenchmarkId::new("many", n), &n, |b, &_n| {
-      b.iter_batched(|| net.clone(), |mut net| black_box(net.normal(&book)), criterion::BatchSize::SmallInput);
-    });
+fn run_dir(dir: &PathBuf, group: Option<&str>, c: &mut Criterion) {
+  let files = std::fs::read_dir(dir).unwrap();
+
+  for file in files.flatten() {
+    let file_path = &file.path();
+    let file_name = file_path.file_stem().unwrap().to_string_lossy();
+
+    let Some(ext) = file_path.extension() else {
+      if file_path.is_dir() {
+        match group {
+          Some(group) => run_dir(file_path, Some(&format!("{group}/{file_name}")), c),
+          None => run_dir(file_path, Some(file_name.as_ref()), c),
+        };
+      }
+      continue;
+    };
+
+    let (book, net) = if ext == "hvmc" {
+      load_from_core(file_path)
+    } else if ext == "hvm" {
+      load_from_lang(file_path)
+    } else {
+      panic!("invalid file found: {}", file_path.to_string_lossy())
+    };
+
+    match group {
+      Some(group) => benchmark_group(&file_name, group, book, net, c),
+      None => benchmark(&file_name, book, net, c),
+    }
   }
 }
 
-fn fusion_benchmark(c: &mut Criterion) {
-  let mut group = c.benchmark_group("fusion");
-  let (book, net) = load_from_lang("./benches/programs/neg_fusion.hvm", 1 << 8, None);
-  group.bench_function(criterion::BenchmarkId::new("neg", "256"), |b| {
-    b.iter_batched(|| net.clone(), |mut net| black_box(net.normal(&book)), criterion::BatchSize::SmallInput);
+fn benchmark(file_name: &str, book: run::Book, net: run::Net, c: &mut Criterion) {
+  c.bench_function(file_name, |b| {
+    b.iter_batched(
+      || net.clone(),
+      |net| black_box(black_box(net).normal(black_box(&book))),
+      criterion::BatchSize::SmallInput,
+    );
+  });
+}
+
+fn benchmark_group(file_name: &str, group: &str, book: run::Book, net: run::Net, c: &mut Criterion) {
+  c.benchmark_group(group).bench_function(file_name, |b| {
+    b.iter_batched(
+      || net.clone(),
+      |net| black_box(black_box(net).normal(black_box(&book))),
+      criterion::BatchSize::SmallInput,
```

Comment:
this should probably be experimentally confirmed, but maybe we should do something like
```suggestion
      match net.heap.data.len() {
        0..(1 << 8)               => criterion::BatchSize::SmallInput,
        (1 << 8) ..(1 << 16) => criterion::BatchSize::LargeInput,
        (1 << 16)..               => criterion::BatchSize::PerIteration,
      }
```

take a look at `BatchSize` in criterion, and run it in your machine and see if it runs faster/doesn't crash your PC

Reply:
I did not notice any significant difference using this on my machine. Should we have it like that anyway?